plc4go/internal/ads/Discoverer.go

```diff
@@ -0,0 +1,250 @@
+					opts["targetAmsPort"] = []string{"851"}
+
+					attributes := make(map[string]values.PlcValue)
+					attributes["hostName"] = values2.NewPlcSTRING(hostNameBlock.GetHostName().GetText())
+					if versionBlock != nil {
+						versionData := versionBlock.GetVersionData()
+						patchVersion := (int(versionData[3])&0xFF)<<8 | (int(versionData[2]) & 0xFF)
+						attributes["twinCatVersion"] = values2.NewPlcSTRING(fmt.Sprintf("%d.%d.%d", int(versionData[0])&0xFF, int(versionData[1])&0xFF, patchVersion))
+					}
+					if fingerprintBlock != nil {
+						attributes["fingerprint"] = values2.NewPlcSTRING(string(fingerprintBlock.GetData()))
+					}
+					// TODO: Find out how to handle the OS Data
+
+					// Add an entry to the results.
+					remoteAddress, err2 := url.Parse("udp://" + strconv.Itoa(int(remoteAmsNetId.GetOctet1())) + "." +
+						strconv.Itoa(int(remoteAmsNetId.GetOctet2())) + "." +
+						strconv.Itoa(int(remoteAmsNetId.GetOctet3())) + "." +
+						strconv.Itoa(int(remoteAmsNetId.GetOctet4())) + ":" +
+						strconv.Itoa(int(driverModel.AdsConstants_ADSTCPDEFAULTPORT)))
+					if err2 == nil {
+						plcDiscoveryItem := &internalModel.DefaultPlcDiscoveryItem{
+							ProtocolCode:  "ads",
+							TransportCode: "tcp",
+							TransportUrl:  *remoteAddress,
+							Options:       opts,
+							Name:          hostNameBlock.GetHostName().GetText(),
+							Attributes:    attributes,
+						}
+
+						// Pass the event back to the callback
+						callback(plcDiscoveryItem)
+					}
+				}
+			}
+		}
+	}()
+
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Find out which interfaces to use for sending out search requests
+
+	allInterfaces, err := net.Interfaces()
+	if err != nil {
+		return err
+	}
+
+	// If no device is explicitly selected via option, simply use all of them
+	// However if a discovery option is present to select a device by name, only
+	// add those devices matching any of the given names.
+	var interfaces []net.Interface
+	deviceNames := options.FilterDiscoveryOptionsDeviceName(discoveryOptions)
+	if len(deviceNames) > 0 {
+		for _, curInterface := range allInterfaces {
+			for _, deviceNameOption := range deviceNames {
+				if curInterface.Name == deviceNameOption.GetDeviceName() {
+					interfaces = append(interfaces, curInterface)
+					break
+				}
+			}
+		}
+	} else {
+		interfaces = allInterfaces
+	}
+
+	////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+	// Send out search requests on all selected interfaces
+
+	// Iterate over all network devices of this system.
+	for _, interf := range interfaces {
+		addrs, err := interf.Addrs()
+		if err != nil {
+			return err
+		}
+		// Iterate over all addresses the current interface has configured
+		// For ADS we're only interested in IPv4 addresses, as it doesn't
+		// seem to work with IPv6.
+		for _, addr := range addrs {
```

Comment:
In C-Bus discovery I moved those inner loops in functions, maybe that would help a bit here